File: src/3640.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def maxSumTrionic(self, nums: List[int]) -> int:
        retVal = float("-inf")

        numsSize = len(nums)

        i = 0
        while i < numsSize:
            j = i + 1
            res = 0

            # first segment: increasing segment
            while (j < numsSize) and (nums[j - 1] < nums[j]):
                j += 1
            p = j - 1
            if p == i:
                i += 1
                continue

            # second segment: decreasing segment
            res += (nums[p] + nums[p - 1])
            while (j < numsSize) and (nums[j - 1] > nums[j]):
                res += nums[j]
                j += 1
            q = j - 1
            if (q == p) or (q == numsSize - 1) or (j < numsSize and nums[j] <= nums[q]):
                i = q
                continue

            # third segment: increasing segment
            res += nums[q + 1]

            # find the maximum sum of the third segment
            maxSum = 0
            currrentSum = 0
            k = q + 2
            while (k < numsSize) and (nums[k] > nums[k - 1]):
                currrentSum += nums[k]
                maxSum = max(maxSum, currrentSum)
                k += 1
            res += maxSum

            # find the maximum sum of the first segment
            maxSum = 0
            currrentSum = 0
            for k in range(p - 2, i - 1, -1):
                currrentSum += nums[k]
                maxSum = max(maxSum, currrentSum)
            res += maxSum

            retVal = max(retVal, res)
            i = q

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for nums in [[0, -2, -1, -3, 0, 2, -1], [1, 4, 2, 7]]:
            # /* Example
            #  *  Input: nums = [0,-2,-1,-3,0,2,-1]
            #  *  Output: -4
            #  *
            #  *  Input: nums = [1,4,2,7]
            #  *  Output: 14
            #  */
            logging.info("Input: nums = %s", nums)

            retVal = pSolution.maxSumTrionic(nums)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)

File: src/3640.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <vector>

using namespace std;

class Solution {
   public:
    long long maxSumTrionic(vector<int>& nums) {
        long long retVal = numeric_limits<long long>::min();

        int numsSize = nums.size();

        int p, q, j;
        long long maxSum, currrentSum, res;
        for (int i = 0; i < numsSize; i++) {
            j = i + 1;
            res = 0;

            // first segment
            for (; ((j < numsSize) && (nums[j - 1] < nums[j])); j++);
            p = j - 1;
            if (p == i) {
                continue;
            }

            // second segment
            res += (nums[p] + nums[p - 1]);
            for (; ((j < numsSize) && (nums[j - 1] > nums[j])); j++) {
                res += nums[j];
            }
            q = j - 1;
            if (q == p || q == numsSize - 1 || (nums[j] <= nums[q])) {
                i = q;
                continue;
            }

            // third segment
            res += nums[q + 1];

            // find the maximum sum of the third segment
            maxSum = 0;
            currrentSum = 0;
            for (int k = q + 2; k < numsSize && nums[k] > nums[k - 1]; k++) {
                currrentSum += nums[k];
                maxSum = max(maxSum, currrentSum);
            }
            res += maxSum;

            // find the maximum sum of the first segment
            maxSum = 0;
            currrentSum = 0;
            for (int k = p - 2; k >= i; k--) {
                currrentSum += nums[k];
                maxSum = max(maxSum, currrentSum);
            }
            res += maxSum;

            // update answer
            retVal = max(retVal, res);
            i = q - 1;
        }

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        vector<int> nums;
    };
    vector<subject> testData{{{0, -2, -1, -3, 0, 2, -1}}, {{1, 4, 2, 7}}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: nums = [0,-2,-1,-3,0,2,-1]
     *  Output: -4
     *
     *  Input: nums = [1,4,2,7]
     *  Output: 14
     */

    Solution cSolution;
    long long answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: nums = [";
        for (long unsigned int j = 0; j < testData[i].nums.size(); ++j) {
            cout << ((j == 0) ? "" : ",") << testData[i].nums[j];
        }
        cout << "]\n";

        answer = cSolution.maxSumTrionic(testData[i].nums);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}
